crates/adapters/src/transport/durable/mod.rs

```diff
@@ -0,0 +1,192 @@
+            ),
+        ])
+    });
+
+impl dyn DurableInputTransport {
+    /// Lookup input transport by `name`, which should be e.g. `kafka` for Kafka
+    /// transport.
+    pub fn get_transport(name: &str) -> Option<&'static dyn DurableInputTransport> {
+        INPUT_TRANSPORT.get(name).map(|f| &**f)
+    }
+}
+
+/// Factory to produce [`DurableInputEndpoint`]s.
+///
+/// Creates durable input transport endpoints for one kind of transport.
+///
+/// A "durable" data input transport is one that allows the same input to be
+/// retrieved as many times as needed, to support fault tolerance.
+pub trait DurableInputTransport: Send + Sync {
+    /// Unique name of the data transport.
+    fn name(&self) -> Cow<'static, str>;
+
+    /// Creates a new `DurableInputEndpoint`.
+    ///
+    /// Creates and initializes a durable transport endpoint with
+    /// transport-specific configuration `config`.
+    ///
+    /// # Errors
+    ///
+    /// Fails if the specified configuration is invalid or the endpoint failed
+    /// to initialize (e.g., the endpoint was not able to establish a network
+    /// connection).
+    fn new_endpoint(&self, config: &YamlValue) -> AnyResult<Box<dyn DurableInputEndpoint>>;
+}
+
+/// Durable data input transport endpoint created by [`DurableInputTransport`].
+///
+/// A "durable" data input transport divides input into steps numbered
+/// sequentially starting from zero.  Each step has durable content; that is, if
+/// it is read again, it will be the same as the first time.
+pub trait DurableInputEndpoint {
+    /// The range of steps that are available to be read.
+    fn steps(&self) -> AnyResult<Range<Step>>;
+
+    /// Opens the endpoint for reading steps and possibly adding them at the
+    /// end.  The reader passes data and status information to `consumer`.
+    fn open(&self, consumer: Box<dyn DurableInputConsumer>) -> Box<dyn DurableInputReader>;
+
+    /// Notifies the endpoint that steps less than `step` aren't needed
+    /// anymore.  It may optionally discard them.
+    fn expire(&self, _step: Step) {}
+}
+
+/// Reader and appender for [`DurableInputTransport`].
+///
+/// A `DurableInputReader` reads from a durable input transport step-by-step.
+/// It first reads the step specified in the first call to [`next`](Self::next),
+/// and advances as far as the highest step passed in any later call to
+/// [`next`](Self::next).  A reader can also cause new steps to be appended, by
+/// calling [`next`](Self::next) with a step number that does not yet exist.
+///
+/// Call [`DurableInputEndpoint::open`] to create a `DurableInputReader`.  The
+/// reader passes the data that it reads to the [consumer](DurableInputConsumer)
+/// passed to the `open` call.
+///
+/// The reader reads and appends asynchronously in separate threads.  Drop the
+/// reader to stop these threads and stop passing data to the consumer.
+///
+/// > 🚨️**Warning**: Only a single appender should be active at once for a given
```

Comment:
wasn't obvious to me why this is warning, maybe explain why it's bad if there is more than one

Reply:
Well, otherwise we need to have some sort of consensus among all the active appenders to figure out which one should actually do it. Kafka doesn't provide the right primitives for that so we'd have to adopt or implement one of our own. Maybe in the end we'll need to do that anyway--and then we can remove this warning--but I don't want to start there.

I folded this in:

```patch
diff --git a/crates/adapters/src/transport/durable/mod.rs b/crates/adapters/src/transport/durable/mod.rs
index 975434a2..cdde88cb 100644
--- a/crates/adapters/src/transport/durable/mod.rs
+++ b/crates/adapters/src/transport/durable/mod.rs
@@ -112,7 +112,8 @@ pub trait DurableInputEndpoint {
 /// reader to stop these threads and stop passing data to the consumer.
 ///
 /// > 🚨️**Warning**: Only a single appender should be active at once for a given
-/// endpoint.
+/// endpoint to avoid races from separate instances trying to add data at the
+/// same time.
 pub trait DurableInputReader {
     /// Requests that the endpoint reads steps up to and including `step`.  If
     /// this is the first call for a given reader, then `step` will be the first
```